File: apps/yryvu-app/src/ipc/integrations/create.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

import type { IssueDetail } from "./issues";

/// Cross-provider input for creating a pull / merge request. Source
/// + target branches are required everywhere; draft state is honored
/// by GitHub + GitLab (encoded via title prefix), Gitea route is
/// `NotImplemented` until Gitea PR detail lands. `labels`,
/// `assignees`, `reviewers` and `milestone` are arrays of opaque
/// `Identifier.id` strings the user picked from the dropdowns.
export interface CreatePrInput {
  title: string;
  body?: string;
  headRef: string;
  baseRef: string;
  draft?: boolean;
  labels?: string[];
  assignees?: string[];
  reviewers?: string[];
  milestone?: string;
}

/// Create a new pull / merge request on `owner/repo`. Returns the
/// freshly created `PullRequestDetail`. Imported lazily via the same
/// module barrel that exposes `PullRequestDetail`.
export function integrationCreatePr(
  integrationType: string,
  owner: string,
  repo: string,
  input: CreatePrInput,
): Promise<import("../pr_detail").PullRequestDetail> {
  return invoke<import("../pr_detail").PullRequestDetail>(
    "integration_create_pr",
    {
      integrationType,
      owner,
      repo,
      input,
    },
  );
}

/// Cross-provider input for creating an issue. `title` required;
/// `body` markdown + selector arrays optional. `labels`, `assignees`
/// and `milestone` carry opaque `Identifier.id` strings the user
/// picked from the dropdowns.
export interface CreateIssueInput {
  title: string;
  body?: string;
  labels?: string[];
  assignees?: string[];
  milestone?: string;
}

/// Create a new issue on `owner/repo`. Returns the freshly created
/// `IssueDetail`, ready to route into the detail panel without a
/// follow-up fetch.
export function integrationCreateIssue(
  integrationType: string,
  owner: string,
  repo: string,
  input: CreateIssueInput,
): Promise<IssueDetail> {
  return invoke<IssueDetail>("integration_create_issue", {
    integrationType,
    owner,
    repo,
    input,
  });
}

File: crates/yryvu-bridge/src/commands/integrations/create.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::Path;

use tauri::AppHandle;

use crate::integrations::{self, CreateIssueInput, CreatePrInput, IssueDetail, PullRequestDetail};

use super::super::integration_routing::is_self_hosted;
use super::sidecar_path;

/// Create a new pull / merge request on `owner/repo` via the named
/// provider. Same auth + hostname routing as the other PR commands;
/// returns the freshly created [`PullRequestDetail`] so the frontend
/// can route into the detail panel without a follow-up GET.
#[tauri::command]
pub async fn integration_create_pr(
    app: AppHandle,
    profile_id: Option<String>,
    integration_type: String,
    owner: String,
    repo: String,
    input: CreatePrInput,
) -> Result<PullRequestDetail, String> {
    let path = sidecar_path(&app)?;
    let auth = tauri::async_runtime::spawn_blocking({
        let integration_type = integration_type.clone();
        let profile_id = profile_id.clone();
        move || {
            integrations::get_integration(&path as &Path, profile_id.as_deref(), &integration_type)
        }
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?
    .ok_or_else(|| format!("integration '{integration_type}' is not connected"))?;
    let hostname = is_self_hosted(&integration_type)
        .then_some(auth.hostname.as_deref())
        .flatten();
    integrations::create_pr(
        &integration_type,
        &auth.token,
        hostname,
        &owner,
        &repo,
        &input,
    )
    .await
    .map_err(|e| e.to_string())
}

/// Create a new issue on `owner/repo` via the named provider. Same
/// auth + hostname routing as the other issue commands; returns the
/// freshly created [`IssueDetail`] so the frontend can route to the
/// detail panel without a follow-up GET.
#[tauri::command]
pub async fn integration_create_issue(
    app: AppHandle,
    profile_id: Option<String>,
    integration_type: String,
    owner: String,
    repo: String,
    input: CreateIssueInput,
) -> Result<IssueDetail, String> {
    let path = sidecar_path(&app)?;
    let auth = tauri::async_runtime::spawn_blocking({
        let integration_type = integration_type.clone();
        let profile_id = profile_id.clone();
        move || {
            integrations::get_integration(&path as &Path, profile_id.as_deref(), &integration_type)
        }
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?
    .ok_or_else(|| format!("integration '{integration_type}' is not connected"))?;
    let hostname = is_self_hosted(&integration_type)
        .then_some(auth.hostname.as_deref())
        .flatten();
    integrations::create_issue(
        &integration_type,
        &auth.token,
        hostname,
        &owner,
        &repo,
        &input,
    )
    .await
    .map_err(|e| e.to_string())
}
